### ISSUE TITLE
ColumnToggler: Inaccessible for screen-readers and voice commands

### ISSUE THREAD

**Issue by MarkusChur** [primefaces/primefaces#4979]
# Summary
The ColumnToggler component is not correctly read out by screen-readers. It is also not accessible by voice commands of speech recognition software. 
- There is no relation between checkbox and label
  - The hidden `<input>` does not have an `id` attribute
  - The `<label>` does not have a `for` attribute
- The `<div>` that represents the checkbox doesn't have a `role` attribute
- The `<div>` that represents the checkbox doesn't have `aria-checked` attribute

Screen-reader and voice command issue are combined here because the fix for this the same.
I've used NVDA and Dragon for verification.

## 1) Environment
- Does not work with 7.0.4
- Does not work with the showcase
https://www.primefaces.org/showcase/ui/data/datatable/columnToggler.xhtml

## 2) Expected behavior - Screen-Reader
Consider the Year column checkbox of the showcase example
Text read out on focus: 
> Year checkbox checked

Text read out on toggle:
> not checked

Text read out on toggle:
> checked

## 3) Actual behavior - Screen-Reader
Consider the Year column checkbox of the showcase example
Text read out on focus: 
> checkbox checked

Text read out on toggle:
> not checked

Text read out on toggle:
> checked

As one can see you are not able to distinguish between checkboxes when you rely on the screen reader.
## 4) Expected behavior - Voice command
Checkbox is listed when calling a command to interact with checkboxes
> Click checkbox

## 5) Actual behavior -  Voice command
UI Elements are not recognized as checkboxes, therefore nothing happens. You are not able to interact with checkboxes

## 6) Steps to reproduce
- Use a screen reader and listen to the generated output of the ColumnToggler
- Try to use voice commands to access the components checkboxes

## 7) Illustration of fixes considering the generated output
One list element of a ColumnToggler checkbox generates the following output:

![Tog1](https://user-images.githubusercontent.com/6985014/60889772-1a174600-a25a-11e9-89c2-6a1659fadc47.PNG)

A fixed version looks like this:
![Tog2](https://user-images.githubusercontent.com/6985014/60889778-1edbfa00-a25a-11e9-9a39-7113a6347123.PNG)

**Comment by melloware** [comment]
These are good suggestions.  I will see if i can fix.

**Comment by melloware** [comment]
Looking at your screenshot i don't see why the DIV should get the role and aria-checked...they should both be for the Input?  I think really the only thing missing is the ID and the FOR="" right?  The screenreader should be looking at those.

**Comment by MarkusChur** [comment]
Just give it a try with a screenreader. I've edited HTML with Chrome developer tools until it was read out correctly. It was necessary to add `role` and `aria-checked` to `<div>` though

**Comment by melloware** [comment]
That seems really strange to me because the DIV is not an input component so why it would be reading that out instead of the same properties on the INPUT doesn't make any sense.  Unless it is because the INPUT is hidden with the ui-helper-hidden-accessible is hiding it from the screenreader because overflow:hidden style???

**Comment by melloware** [comment]
Reading this I think we just need to add tabindex="0" to the INPUT to make it accessible by the Screenreader.

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/checkbox_role

**Comment by MarkusChur** [comment]
I think touching tabindex is not the way to go. Take a look at accessible examples for checkboxes. Adding attributes to the `<div>` is the way to go.
https://www.w3.org/TR/wai-aria-practices/examples/checkbox/checkbox-1/checkbox-1.html

**Comment by MarkusChur** [comment]
Verified that it's not working correctly with `opacity: 0` as well. It could be that `aria-checked` causes a bug within NVDA when set on `<input type="checkbox">`.
You are definitely right that setting aria attributes on the `<div>` makes no sense, unless the `<input>` is not interpreted correctly.
I'll investigate tomorrow and let you know if I find a better solution.

**Comment by melloware** [comment]
thanks I just want to make sure we make the "right" fix.

**Comment by MarkusChur** [comment]
Here are the results:
# Screen Reader
By fixing the relation of helper `<input>` and `<label>` with `id` attribute, the checkboxes are correctly read out with 7.1-SNAPSHOT.
By refactoring the checkbox to something similar than  https://www.w3.org/TR/wai-aria-practices/examples/checkbox/checkbox-1/checkbox-1.html without `<input>` also fixes the issue.

# Voice Commands
The `<input>` is not recognized by voice commands. I got it recognized by moving it next to the label and by removing the `<div class="ui-helper-hidden-accessible">`. In all other cases it does not react to voice commands. The `<input>` can then be visually hidden by adding a style class with `position:absolute; opacity:0;`.
The `<input>` is not necessary right? It could all be done similar to https://www.w3.org/TR/wai-aria-practices/examples/checkbox/checkbox-1/checkbox-1.html
By doing so the checkboxes would be recognized by voice commands.
If you want to keep the `<input>` then it has to be moved and restyled. All resizing has to be removed, in my tests  `box-sizing` as well. 

# Conclusion
Fixing both interferes. In my opinion the cleanest solution is to remove the `<input>` and refactor it to be like https://www.w3.org/TR/wai-aria-practices/examples/checkbox/checkbox-1/checkbox-1.html

**Comment by melloware** [comment]
I would like @mertsincan to weigh in on this before making any changes.

**Comment by MarkusChur** [comment]
I have a working fix which I'm eager to provide if you are open for such a refactoring.

**Comment by melloware** [comment]
Sure submit a PR and we can review it.
